File: 01-sistema-gestion-datos/aplicacion/casos_uso/caso_uso_crear_usuario.py
"""
Caso de Uso: Crear Usuario
Implementa la lógica de negocio para crear un nuevo usuario
"""
from typing import Optional
from dataclasses import dataclass
from dominio.entidades.usuario import Usuario
from dominio.value_objects.email import Email
from dominio.value_objects.nombre_usuario import NombreUsuario
from dominio.value_objects.contraseña import Contraseña
from dominio.interfaces.repositorio_usuario import RepositorioUsuario
from dominio.servicios.servicio_autenticacion import ServicioAutenticacion
from aplicacion.dto.crear_usuario_dto import CrearUsuarioDTO
from aplicacion.dto.usuario_dto import UsuarioDTO
from aplicacion.excepciones.excepciones_aplicacion import (
    UsuarioYaExisteError,
    EmailYaExisteError,
    NombreUsuarioYaExisteError,
    ErrorValidacionError
)


@dataclass
class CasoUsoCrearUsuario:
    """
    Caso de uso para crear un nuevo usuario
    Implementa el patrón Use Case de Clean Architecture
    """
    
    repositorio_usuario: RepositorioUsuario
    servicio_autenticacion: ServicioAutenticacion
    
    async def ejecutar(self, datos_usuario: CrearUsuarioDTO) -> UsuarioDTO:
        """
        Ejecutar el caso de uso para crear usuario
        
        Args:
            datos_usuario: Datos del usuario a crear
            
        Returns:
            UsuarioDTO del usuario creado
            
        Raises:
            UsuarioYaExisteError: Si el usuario ya existe
            EmailYaExisteError: Si el email ya está en uso
            NombreUsuarioYaExisteError: Si el nombre de usuario ya está en uso
            ErrorValidacionError: Si los datos no son válidos
        """
        # Validar datos de entrada
        await self._validar_datos_entrada(datos_usuario)
        
        # Crear value objects
        email = Email(datos_usuario.email)
        nombre_usuario = NombreUsuario(datos_usuario.nombre_usuario)
        contraseña = Contraseña(datos_usuario.contraseña)
        
        # Verificar que no exista un usuario con el mismo email
        if await self.repositorio_usuario.existe_email(email):
            raise EmailYaExisteError(f"Ya existe un usuario con el email: {email}")
        
        # Verificar que no exista un usuario con el mismo nombre de usuario
        if await self.repositorio_usuario.existe_nombre_usuario(nombre_usuario):
            raise NombreUsuarioYaExisteError(
                f"Ya existe un usuario con el nombre: {nombre_usuario}"
            )
        
        # Crear entidad usuario
        usuario = Usuario(
            email=email,
            nombre_usuario=nombre_usuario,
            contraseña=contraseña,
            nombre_completo=datos_usuario.nombre_completo,
            biografia=datos_usuario.biografia,
            esta_activo=True,
            es_superusuario=False
        )
        
        # Guardar usuario en el repositorio
        usuario_creado = await self.repositorio_usuario.guardar(usuario)
        
        # Convertir a DTO y retornar
        return UsuarioDTO.desde_entidad(usuario_creado)
    
    async def _validar_datos_entrada(self, datos_usuario: CrearUsuarioDTO) -> None:
        """
        Validar los datos de entrada del usuario
        
        Args:
            datos_usuario: Datos del usuario a validar
            
        Raises:
            ErrorValidacionError: Si los datos no son válidos
        """
        errores = []
        
        # Validar email
        if not datos_usuario.email or not datos_usuario.email.strip():
            errores.append("El email es requerido")
        
        # Validar nombre de usuario
        if not datos_usuario.nombre_usuario or not datos_usuario.nombre_usuario.strip():
            errores.append("El nombre de usuario es requerido")
        
        # Validar contraseña
        if not datos_usuario.contraseña or not datos_usuario.contraseña.strip():
            errores.append("La contraseña es requerida")
        
        # Validar nombre completo (opcional pero si se proporciona debe ser válido)
        if datos_usuario.nombre_completo and len(datos_usuario.nombre_completo.strip()) < 2:
            errores.append("El nombre completo debe tener al menos 2 caracteres")
        
        # Validar biografía (opcional pero si se proporciona debe ser válida)
        if datos_usuario.biografia and len(datos_usuario.biografia.strip()) > 500:
            errores.append("La biografía no puede exceder 500 caracteres")
        
        if errores:
            raise ErrorValidacionError(f"Errores de validación: {'; '.join(errores)}")

File: 06-arquitectura-microservicios-java/microservicios/servicio-usuarios/src/main/java/com/portfolio/arquitectura/usuarios/aplicacion/casosUso/CasoUsoCrearUsuario.java
package com.portfolio.arquitectura.usuarios.aplicacion.casosUso;

import com.portfolio.arquitectura.usuarios.dominio.entidad.Usuario;
import com.portfolio.arquitectura.usuarios.dominio.entidad.RolUsuario;
import com.portfolio.arquitectura.usuarios.dominio.entidad.EstadoUsuario;
import com.portfolio.arquitectura.usuarios.dominio.repositorio.RepositorioUsuario;
import com.portfolio.arquitectura.usuarios.dominio.objetoValor.CorreoElectronico;
import com.portfolio.arquitectura.usuarios.aplicacion.dto.SolicitudCrearUsuario;
import com.portfolio.arquitectura.usuarios.aplicacion.dto.RespuestaUsuario;
import com.portfolio.arquitectura.usuarios.aplicacion.mapeador.MapeadorUsuario;
import com.portfolio.arquitectura.usuarios.aplicacion.excepcion.UsuarioYaExisteExcepcion;
import com.portfolio.arquitectura.usuarios.aplicacion.excepcion.DatosUsuarioInvalidosExcepcion;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Caso de Uso para Crear un Nuevo Usuario
 * 
 * Implementa la lógica de negocio para la creación de usuarios,
 * incluyendo validaciones, encriptación de contraseña y notificaciones.
 * 
 * @author Gabriel - Arquitecto de Software
 */
@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class CasoUsoCrearUsuario {

    private final RepositorioUsuario repositorioUsuario;
    private final PasswordEncoder codificadorContrasena;
    private final MapeadorUsuario mapeadorUsuario;
    private final PublicadorEventoUsuario publicadorEventoUsuario;

    /**
     * Crea un nuevo usuario en el sistema
     * 
     * @param solicitud Datos del usuario a crear
     * @return Usuario creado
     * @throws UsuarioYaExisteExcepcion si el usuario ya existe
     * @throws DatosUsuarioInvalidosExcepcion si los datos son inválidos
     */
    public RespuestaUsuario ejecutar(SolicitudCrearUsuario solicitud) {
        log.info("Iniciando creación de usuario con correo: {}", solicitud.getCorreoElectronico());

        // Validar datos de entrada
        validarSolicitud(solicitud);

        // Crear correo electrónico objeto valor
        CorreoElectronico correo = CorreoElectronico.de(solicitud.getCorreoElectronico());

        // Verificar si el usuario ya existe
        if (repositorioUsuario.existePorCorreo(correo)) {
            log.warn("Intento de crear usuario con correo existente: {}", solicitud.getCorreoElectronico());
            throw new UsuarioYaExisteExcepcion("Ya existe un usuario con el correo: " + solicitud.getCorreoElectronico());
        }

        if (repositorioUsuario.existePorNombreUsuario(solicitud.getNombreUsuario())) {
            log.warn("Intento de crear usuario con nombre de usuario existente: {}", solicitud.getNombreUsuario());
            throw new UsuarioYaExisteExcepcion("Ya existe un usuario con el nombre de usuario: " + solicitud.getNombreUsuario());
        }

        // Crear entidad usuario
        Usuario usuario = Usuario.builder()
                .correoElectronico(correo.getValor())
                .nombreUsuario(solicitud.getNombreUsuario())
                .contrasena(codificadorContrasena.encode(solicitud.getContrasena()))
                .nombre(solicitud.getNombre())
                .apellido(solicitud.getApellido())
                .numeroTelefono(solicitud.getNumeroTelefono())
                .estado(EstadoUsuario.PENDIENTE_VERIFICACION)
                .rol(determinarRolUsuario(solicitud))
                .correoVerificado(false)
                .build();

        // Guardar usuario
        Usuario usuarioGuardado = repositorioUsuario.guardar(usuario);
        log.info("Usuario creado exitosamente con ID: {}", usuarioGuardado.getId());

        // Publicar evento de usuario creado
        publicadorEventoUsuario.publicarUsuarioCreado(usuarioGuardado);

        // Enviar correo de verificación
        publicadorEventoUsuario.publicarSolicitudVerificacionCorreo(usuarioGuardado);

        return mapeadorUsuario.aRespuesta(usuarioGuardado);
    }

    /**
     * Valida los datos de entrada
     */
    private void validarSolicitud(SolicitudCrearUsuario solicitud) {
        if (solicitud == null) {
            throw new DatosUsuarioInvalidosExcepcion("Solicitud no puede ser nula");
        }

        if (solicitud.getCorreoElectronico() == null || solicitud.getCorreoElectronico().trim().isEmpty()) {
            throw new DatosUsuarioInvalidosExcepcion("Correo electrónico es requerido");
        }

        if (solicitud.getNombreUsuario() == null || solicitud.getNombreUsuario().trim().isEmpty()) {
            throw new DatosUsuarioInvalidosExcepcion("Nombre de usuario es requerido");
        }

        if (solicitud.getContrasena() == null || solicitud.getContrasena().length() < 8) {
            throw new DatosUsuarioInvalidosExcepcion("Contraseña debe tener al menos 8 caracteres");
        }

        if (solicitud.getNombre() == null || solicitud.getNombre().trim().isEmpty()) {
            throw new DatosUsuarioInvalidosExcepcion("Nombre es requerido");
        }

        if (solicitud.getApellido() == null || solicitud.getApellido().trim().isEmpty()) {
            throw new DatosUsuarioInvalidosExcepcion("Apellido es requerido");
        }

        // Validar formato de nombre de usuario
        if (!solicitud.getNombreUsuario().matches("^[a-zA-Z0-9_]{3,20}$")) {
            throw new DatosUsuarioInvalidosExcepcion("Nombre de usuario debe contener solo letras, números y guiones bajos, entre 3 y 20 caracteres");
        }
    }

    /**
     * Determina el rol del usuario basado en el contexto
     */
    private RolUsuario determinarRolUsuario(SolicitudCrearUsuario solicitud) {
        // Por defecto, todos los usuarios son CLIENTE
        // En un sistema real, esto podría basarse en:
        // - Código de invitación
        // - Dominio de correo electrónico
        // - Configuración del sistema
        return RolUsuario.CLIENTE;
    }
}
